server/src/models/gorilla.rs

```diff
@@ -131,106 +127,63 @@ impl Gorilla {
     }
 }
 
-/// Compute the minimum value for a time series segment whose values are
+/// Compute the sum of the values for a time series segment whose values are
 /// compressed using Gorilla's compression method for floating-point values.
-pub fn min(
-    start_time: Timestamp,
-    end_time: Timestamp,
-    sampling_interval: i32,
-    model: &[u8],
-) -> Value {
-    let decompressed_values =
-        decompress_values_to_array(start_time, end_time, sampling_interval, model);
-    aggregate::min(&decompressed_values).unwrap()
-}
+pub fn sum(start_time: Timestamp, end_time: Timestamp, timestamps: &[u8], values: &[u8]) -> Value {
+    let length = models::length(start_time, end_time, timestamps);
+    let mut bits = BitReader::try_new(values).unwrap();
+    let mut leading_zeros = u8::MAX;
+    let mut trailing_zeros: u8 = 0;
+    let mut last_value = bits.read_bits(models::VALUE_SIZE_IN_BITS);
 
-/// Compute the maximum value for a time series segment whose values are
-/// compressed using Gorilla's compression method for floating-point values.
-pub fn max(
-    start_time: Timestamp,
-    end_time: Timestamp,
-    sampling_interval: i32,
-    model: &[u8],
-) -> Value {
-    let decompressed_values =
-        decompress_values_to_array(start_time, end_time, sampling_interval, model);
-    aggregate::max(&decompressed_values).unwrap()
-}
+    // The first value is stored uncompressed using size_of::<Value> bits.
+    let mut sum = Value::from_bits(last_value);
 
-/// Compute the sum of the values for a time series segment whose values are
-/// compressed using Gorilla's compression method for floating-point values.
-pub fn sum(
-    start_time: Timestamp,
-    end_time: Timestamp,
-    sampling_interval: i32,
-    model: &[u8],
-) -> Value {
-    let decompressed_values =
-        decompress_values_to_array(start_time, end_time, sampling_interval, model);
-    aggregate::sum(&decompressed_values).unwrap()
-}
+    // Then values are stored using XOR and a variable length binary encoding.
```

Comment:
The code is purposely replicated as `sum()` does not need to allocate arrays for storing the time series ids, timestamps, and values while `grid()` does. As the code to compute the sum or add the values to the builders must be inserted into the main loop, I could not think of a good way to efficiently move the shared code to a separate function. So since the two function are right next to each other, replicating the small amount of code in each function seemed like the least unpleasant solution. I have added a comment to `sum()` to document why it contains code from `grid()`. I have also added comments to both functions stating that they must both be updated with any changed to the decompression.

Reply:
Makes sense.